Open Source as a Service

stefwalter opened the pull request:
This is an article describing why and what it means to enable Open Source contributions to a Service. 

Sharing the source code to a Service is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.

stefwalter committed:
WIP complete the article basics

stefwalter commented:
@martinpitt @thrix @TomasTomecek @major @durandom @emmanuelbernard @gsaslis Interested in reviewing? Here's an introductory blog post about the Open Source Services topic we've talked about.

I'm thinking of following it up with more content, perhaps some of the playbook ideas we've already drafted together. But one thing at a time.

martinpitt commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,155 @@
+Title: Your Service is not Open Source
+Date: 2018-02-06 18:35
```
Do you have your time travel permit from the Ministry of Truth? :grin:

stefwalter replied:
Heh. Fixed.

martinpitt commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,155 @@
+first deploy and operate their own clone: Something that they explicitly
+chose not to do when they elected to use the software as a service.
+
+Even if the user becomes an operator, and deploys a such clone, it
+is devoid of their own data, devoid of other users, of connections
+to other services, of infrastructure scalability, pipelines, monitoring,
+backups, ... you name it.
```
This is a bit harsh IMHO. First, the whole network of users, credentials, data etc. are the actual and precious assets of the "primary" SaaS, and exposing/exporting all of that is not only undesirable and inappropriate, it's outright illegal (data protection, private projects, exposing secrets, etc.). Also the whole point of hosting your own instance is that you can work with your own data, test it, and break it.

The second half (scalability, pipelines, monitoring, backups) just means that the deployment scripts are incomplete -- things like metrics collection or database replication should be part of the deployment; of course the "other" side of that (pagers, email etc.) is highly specific to a particular user/company. Also, for test instances you most likely don't want that extra overhead anyway.

I understand that this is primarily an argument for *not* hosting your own instance in this context, and for that it's valid -- but that's already handled in the first paragraph.

stefwalter replied:
Good points. I've tried to rewrite this and stick to objective claims. WDYT? Do you feel these paragraphs are more coherent now? I'd be happy for further critique, in areas where my subjectivity remains.

martinpitt commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,155 @@
+
+You still with me?
+
+Yes, that means opening a pull request against a running service
+and running that change before it's merged.
```
Connecting that to your point above -- the *only* way that this is feasible is if your own little shard of that SaaS with your PR operates on a cloned subset of the production data. If any random Joe Developer could submit a PR against gitlab that runs on the production data, what trust could you still place in the gitlab Saas? IMHO, it would be "none".

But again, I think having a playground with empty (or cloned data that belongs to you) is actually preferable, as it gives you the possibility of doing thorough, destructive, and risky tests.

stefwalter replied:
I'll cover this in follow up articles and playbook. I believe that your statements are just as true of team contributions.
Many teams only merge code once they have gotten it into production. See the GitHub internal contribution process.

But yes, I'm happy leaving this as a controvesial topic to be discussed further.

gsaslis replied:
@martinpitt i think the crucial bit that is mentioned above is that this change would only affect your data / how only you consume the service. 

I think of it like this: 
many SaaS offer an option to try their "beta" features or become part of their "fast" channel, with latest changes not-yet fully tested on everyone else. Other SaaS A/B test features all the time, so it's not like all the users of the service experience the same service anyway. 

Switching to "contributor mode" for an Open Source Service would be very similar to those cases but the features you are beta testing are simply your own.

martinpitt commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,155 @@
+
+# Why should I care?
+
+I've been a part of Open Source for over 20 years now. Although I'm
+just one contributor, together we've changed something fundamental
+about the world. Our lives are too short, and software too complex,
+to have one company or one individual invent everything from scratch.
+So we work together across humanity to accomplish what no
+individual team could. This is our legacy, and it has become
+common place.
```
I had to shed a tear of joy here :joy: Well written!

stefwalter replied:
Thanks!

martinpitt commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,155 @@
+Whereas
+[Facebook rapidly deploys changes](https://engineering.fb.com/2017/08/31/web/rapid-release-at-massive-scale/)
+to a few canary machines and scales each change up to production.
+Everyone has their own version of continuous delivery. It's hard to
+take an engineering team seriously if they haven't figured out how
+their team members can get a change rapidly deployed.
+
+As Open Source, I don't see us having pulled this off. Not yet.
```
This is something different, though, so IMHO this breaks the flow of argument here. This paragraph is about the engineers of the SaaS installation doing CD of their own code to their own official installation. I don't know how gitlab or e.g. the Testing Farm cluster handles that, but I can't imagine that CD is so rare in the FOSS world? (FWIW, this is exactly how we handle infra changes in our cockpit CI, even though that is not a public service).

Your plea above was that developers *outside* of that core developer team could submit a PR and get their own shard of the service (*not* get it deployed to the official primary SaaS). The latter is indeed much harder IMHO, but then again I am not aware of any non-open project which does that. In fact, by definition you couldn't even do something like "send a PR against github".

So, I agree that FOSS has not managed to set up this "get your own shard on a PR", and this is indeed something that would give FOSS infra projects a decisive advantage over their closed-source counterparts. But it's certainly not an aspect where FOSS infra projects lag behind closed-source ones. So this could be easy to msunderstand here.

stefwalter replied:
Indeed. This was sloppy of me. I removed a bunch of the "how do we do it" stuff for later posts because the article was getting too long. How does it look now?

martinpitt reviewed (commented):
Very nice article, thanks!

stefwalter committed:
FIXUP changes after feedback from @martinpitt

stefwalter commented:
Thank you @martinpitt for your valuable critique and review. I hope I've addressed the points you raised. 

One point in particular (about sharding/cloning data) is better covered in a follow up, and is a key example of the areas where we must apply existing known deployment techniques to solve.

martinpitt commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+
+ 3. Experience that change when they use the service, or when it acts on their data.
+
+After discussing this with others, I firmly believe we have all the techniques we need,
+whether canary deployments, load balancing, continuous delivery,
+infrastructure as code, and so on. It's been hard to find a single
+problem that an existing practice doesn't solve.
```
Not for your article (perhaps), but a thought: Achieving this would also mean that you have full control of the "real" deployment, and constantly exercise and improve its (re)deployment. The importance of that can hardly be over-estimated for a project's/company's long-term resilience.

martinpitt reviewed (approved):
Thanks, this sounds great to me now! I can't resolve my comments above, but please do so (especially if/when others chime in, to keep track of discussions).

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+a service have explicitly chosen not to be involved in operating
+the software.
+
+If the primary way to use [Postgres was as a Service](https://aws.amazon.com/rds/),
```
"as a Service" ?

(i.e.  add quotes to clarify. This took me a few reads to ensure I understood the point)

durandom replied:
Yes, how about:
"If the primary way how to use Postrgres, would have been to consume it [as a Service](), then that..."

stefwalter replied:
Thanks. Good catch. Updated to use quotes. It matches my flavor better ;)

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+Title: Your Service is not Open Source
+Date: 2020-02-07 18:35
+Tags: open source, technical
+Slug: open-source-as-a-service
```
i think I prefer your "Open Source Services" term. It better communicates intent (open source as a service sounds like you are offering open source itself as a service, while our focus is on services here). It is also a nice play on OSS, but with a very nice & clear differentiation between Software vs. Service.

durandom replied:
arent slugs just `hyphenate($title)` ?

stefwalter replied:
Thanks @gsaslis. I checked that name with some of the others in a call today, and everyone liked "Open Source Services" ... so good call. Will change the slug and start using that name elsewhere as well.

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+Tags: open source, technical
+Slug: open-source-as-a-service
+Category: Technical
+Summary: Sharing the source code to your SaaS is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.
```
How about 👇 ?

Open sourcing the code to your SaaS is insufficient. For a *service* to really be Open Source, we need to effectively enable users to contribute to the SaaS itself.

stefwalter replied:
Better. Thanks.

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+Category: Technical
+Summary: Sharing the source code to your SaaS is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.
+
+Sharing the source code to your SaaS is insufficient to make it
```
"Sharing the source code to" -> "Open Sourcing" ?

stefwalter replied:
Done.

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+Open Source: enabling users to make a change to the software
+they're using.
+
+Some other Open Source services, such as
```
I would change "Some other Open Source services" to save "Open Source Services" as the term to use only for our end game, not for what (gitlab|elastic) currently do. 

"Some other popular services powered by Open Source Software" 

(even though this might now be controversial in elastic's case...)

durandom replied:
agreed. Coining a term is powerful

stefwalter replied:
Agree with coining the term. But I wouldn't like to apply the term to these projects (GitLab is Open Core + Managed Service and ElasticSearch has a dubious license now).

stefwalter replied:
Updated to use the text @gsaslis suggested in quotes.

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+in a way that typical open source projects with a true community
+are resilient to.
+
+I believe that if we enable contributions to services, we reap
```
let's add 

"I believe that if we enable contributions to services, *rather than just the software*,"

stefwalter replied:
Clearer. Thanks.

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+
+In reality the threat here is far more fundamental: Users of services
+cannot readily change something in the service not only because the
+processes is operated by someone else ... but because such users of
```
```suggestion
processes are operated by someone else ... but because such users of
```

gsaslis commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+their team members can get a change rapidly deployed.
+
+As Open Source, we have all the ingredients to have a decisive
+advantage here. To effectively collaborate on services with
```
```suggestion
advantage here. To effectively collaborate on  **Open Source Services**, 
```

gsaslis reviewed (commented):
Thanks for bringing this out @stefwalter ! Love it. 

I've left some minor edits that I think help readability. I also left some comments that I think help bring more emphasis on "Open Source Services" - so that this endeavor has a nice catchy name that will help us pull more people behind it.

radekvokal commented:
Hey @stefwalter , /lgtm . One thing I'd add is to figure out how to capture and open SOPs -- standard operating practices. You're talking about opening access to the components/services and making transparent changes there - that's great but I feel like it's often not about what you run but how you run it - how you operate it. That experience is often encoded in heads of SRE/operational team - debugging and troubleshooting services, actions triggered by monitoring events etc.

durandom commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+Tags: open source, technical
+Slug: open-source-as-a-service
+Category: Technical
+Summary: Sharing the source code to your SaaS is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.
```
We effectively **need to** enable users to contribute to ~~the~~ a running service for it to be Open Source.

stefwalter replied:
Indeed. Thanks for catching that. I replaced this @gsaslis text.

durandom commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+infrastructure as code, and so on. It's been hard to find a single
+problem that an existing practice doesn't solve.
+
+It turns out many engineering teams working have come to that same
```
s/many engineering teams working/many experienced engineering teams//

stefwalter replied:
Thanks.

durandom commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+advantage here. To effectively collaborate on services with
+contributors across humanity.
+
+Lets try to craft a playbook to achieve this basic capability
```
How about

"Lets try to craft a playbook to apply this basic capability that drives Open Source Software to Open Source Services: The ability..."

Using the newly coined term "OSServices"  twice for brainzzz

stefwalter replied:
Awesome. Added.

durandom reviewed (commented):
❤️ it

TomasTomecek commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+Much of the value of a service comes from things other than the code.
+It comes from the infrastructure it runs on, the operational processes,
+the monitoring, the backups, the metrics, the high availability, the
+scalability. It comes from the network effect, with other users, with other
```
```suggestion
scalability, the data, the integration with other tools and services, and more. It comes from the network effect, with other users, with other
```

stefwalter replied:
I think integration with other tools and services is the network effect. So lets add that there.

TomasTomecek commented on content/technical/open-source-as-a-service.md:
```diff
@@ -0,0 +1,153 @@
+You still with me?
+
+Yes, that means opening a pull request against a running service
+and running that change before it's merged.
```
```suggestion
and running that change before it's merged with no effort for the contributor.
```

stefwalter replied:
I'm not sure it's "no effort'. But I changed the second running to be "experiencing" to make it more clear.

TomasTomecek reviewed (approved):
Very nice! Would be nice to add at least a single picture so it doesn't feel like a wall of text :)

stefwalter committed:
FIXUP More changes from @gsaslis

stefwalter committed:
FIXUP from @gsalis content/technical/open-source-as-a-service.md

Co-authored-by: Yorgos Saslis <2420882+gsaslis@users.noreply.github.com>

stefwalter committed:
FIXUP from @gsalis content/technical/open-source-as-a-service.md

Co-authored-by: Yorgos Saslis <2420882+gsaslis@users.noreply.github.com>

stefwalter committed:
FIXUP changes from @durandom

stefwalter committed:
FIXUP from review by @TomasTomecek

stefwalter commented:
Thanks for reviews. I've made all the changes, but I agree we could add a couple pictures. I'll work on those.

@durandom @martinpitt @gsaslis @TomasTomecek @radekvokal Would you be open to me linking back to this pull request at the bottom of the article as a way of thanking you for your feedback?

thrix commented:
I did not find anything to comment, lgtm

martinpitt commented:
@stefwalter : Sure! Linking to PRs is very much the spirit of FOSS :wink:

stefwalter committed:
FIXUP add image to the document

stefwalter committed:
FIXUP Add notice thanking reviewers
